builder/common/ssm/session.go

```diff
@@ -84,21 +86,48 @@ func (s Session) getCommand(ctx context.Context) ([]string, string, error) {
 	return args, *session.SessionId, nil
 }
 
+// terminate an interactive Systems Manager session with a remote instance via the
+// AWS session-manager-plugin. Session cannot be resumed after termination.
+func (s Session) terminateSession(sessionID string, ui packersdk.Ui) {
+	log.Printf("ssm: Terminating PortForwarding session %q", sessionID)
+	_, err := s.SvcClient.TerminateSession(&ssm.TerminateSessionInput{SessionId: aws.String(sessionID)})
+	if err != nil {
+		ui.Error(fmt.Sprintf("Error terminating SSM Session %q, this does not affect the built AMI. Please terminate the session manually: %s", sessionID, err))
+	}
+}
+
 // Start an interactive Systems Manager session with a remote instance via the
 // AWS session-manager-plugin. To terminate the session you must cancell the
 // context. If you do not wish to terminate the session manually: calling
 // StopSession on a instance of this driver will terminate the active session
 // created from calling StartSession.
 // To stop the session you must cancel the context.
 func (s Session) Start(ctx context.Context, ui packersdk.Ui, sessionChan chan struct{}) error {
-	for ctx.Err() == nil {
+	exitSession := false
+	for ctx.Err() == nil && !exitSession {
 		log.Printf("ssm: Starting PortForwarding session to instance %s", s.InstanceID)
 		args, sessionID, err := s.getCommand(ctx)
 		if sessionID != "" {
 			defer func() {
-				_, err := s.SvcClient.TerminateSession(&ssm.TerminateSessionInput{SessionId: aws.String(sessionID)})
-				if err != nil {
-					ui.Error(fmt.Sprintf("Error terminating SSM Session %q. Please terminate the session manually: %s", sessionID, err))
+				s.terminateSession(sessionID, ui)
+			}()
+			// If the instance is terminated the session must be terminated as well
+			// Otherwise the start-session command will exit with status -1
+			go func() {
+				for {
+					instanceState, err := s.Ec2Conn.DescribeInstanceStatus(&ec2.DescribeInstanceStatusInput{
+						InstanceIds: []*string{aws.String(s.InstanceID)},
+					})
+					if err != nil {
+						log.Printf("ssm: Error describing instance status: %s", err)
+					}
+					// if no instance status is returned, the instance is terminated
+					if len(instanceState.InstanceStatuses) == 0 {
+						exitSession = true
+						s.terminateSession(sessionID, ui)
```

Comment:
This should probably not be necessary, as the deferred function already closes the session when we exit `Start`

Reply:
There's one case which I'm not sure how to handle without this:

1. code enters the loop and creates the session this blocks the coroutine when it reaches the `localexec.RunAndStream(cmd, ui, nil)`.
2. the status checker finds out the instance is terminated and set `exitSession` to true.
3. the coroutine is still waiting for the `localexec.RunAndStream(cmd, ui, nil)` to exit. This actually never exits until the command fails and the Bad exit status happens.

Reply:
If we remove the `defer` as pointed out above, we can indeed terminate here, no objection on my part!